contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+
+        #[ink(message)]
+        /// Number of tickets available for sale.
+        ///
+        /// The tickets will be auctioned off one by one.
+        pub fn available_tickets(&self) -> Result<Balance, Error> {
+            self.ticket_balance()
+        }
+
+        #[ink(message)]
+        /// The minimal price the contract allows.
+        pub fn min_price(&self) -> Balance {
+            self.min_price
+        }
+
+        #[ink(message)]
+        /// Address of the reward token contract this contract will accept as payment.
+        pub fn reward_token(&self) -> AccountId {
+            self.reward_token
+        }
+
+        #[ink(message)]
+        /// Address of the ticket token contract this contract will auction off.
+        pub fn ticket_token(&self) -> AccountId {
+            self.ticket_token
+        }
+
+        #[ink(message)]
+        /// Buy one ticket at the current_price.
+        ///
+        /// The caller should make an approval for at least `price()` reward tokens to make sure the
+        /// call will succeed.
+        pub fn buy(&mut self) -> Result<(), Error> {
+            if self.ticket_balance()? > 0 {
+                let price = self.current_price();
+                let account_id = self.env().caller();
+
+                self.take_payment(account_id, price)?;
+                self.give_ticket(account_id)?;
+
+                self.total_proceeds = self.total_proceeds.saturating_add(price);
+                self.tickets_sold = self.tickets_sold.saturating_add(1);
+                self.current_start_block = self.env().block_number();
+                Self::emit_event(self.env(), Event::Buy(Buy { price, account_id }));
+
+                Ok(())
+            } else {
+                Err(Error::MarketplaceEmpty)
+            }
+        }
+
+        #[ink(message)]
+        /// Re-start the auction from the current block.
+        ///
+        /// Note that this will keep the average estimate from previous auctions.
+        pub fn reset(&mut self) -> Result<(), Error> {
+            Self::ensure_role(self.admin())?;
+
+            self.current_start_block = self.env().block_number();
+            Self::emit_event(self.env(), Event::Reset(Reset {}));
+
+            Ok(())
+        }
+
+        fn current_price(&self) -> Balance {
+            let block = self.env().block_number();
+            let elapsed = block.saturating_sub(self.current_start_block.into());
+            self.average_price()
+                .saturating_mul(self.sale_multiplier)
+                .saturating_sub(self.per_block_reduction().saturating_mul(elapsed.into()))
+                .max(self.min_price)
+        }
+
+        fn per_block_reduction(&self) -> Balance {
+            self.average_price()
+                .saturating_div(self.auction_length.into())
+                .max(1u128)
+        }
+
+        fn take_payment(&self, from: AccountId, amount: Balance) -> Result<(), Error> {
+            build_call::<Environment>()
```

Comment:
maybe this could check the allowance before and fail with a meaningful error ?

Reply:
I was basically expecting the other contract to fail with a meaningful variant in that case... In `polkadot.js` all these errors look the same, where do they even show up?

Reply:
to be completely honest I'm not sure how such failure looks like, i.e. whether you get something beyond a revert, possibly not.